### ISSUE TITLE
tbb 4.4.3. appears to be broken with gcc 6.1.0

### ISSUE THREAD

**Issue by davydden** [LLNL/spack#947]
some of the compiler errors are below
```
In file included from /home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/memory:62:0,
                 from ../../include/tbb/tbb_stddef.h:421,
                 from ../../src/tbb/itt_notify.h:24,
                 from ../../src/tbb/itt_notify.cpp:67:
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:243:56: error: macro "min" passed 3 arguments, but takes just 2
     min(const _Tp& __a, const _Tp& __b, _Compare __comp)
                                                        ^
In file included from ../../src/tbb/itt_notify.cpp:47:0:
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:195:5: error: expected unqualified-id before ‘const’
     min(const _Tp& __a, const _Tp& __b)
     ^
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:195:5: error: expected ‘)’ before ‘const’
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:195:5: error: expected initializer before ‘const’
In file included from /home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/memory:62:0,
                 from ../../include/tbb/tbb_stddef.h:421,
                 from ../../src/tbb/itt_notify.h:24,
                 from ../../src/tbb/itt_notify.cpp:67:
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:243:5: error: ‘std::min’ declared as an ‘inline’ variable
     min(const _Tp& __a, const _Tp& __b, _Compare __comp)
     ^~~
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:246:7: error: expected primary-expression before ‘if’
       if (__comp(__b, __a))
       ^~
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:246:7: error: expected ‘}’ before ‘if’
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:246:7: error: expected ‘;’ before ‘if’
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:248:7: error: expected unqualified-id before ‘return’
       return __a;
       ^~~~~~
/home/davydden/spack/opt/spack/linux-x86_64/gcc-4.8/gcc-6.1.0-5vnysvhcqo7ny3t6s2ckjd42eqvkj72h/include/c++/6.1.0/bits/stl_algobase.h:315:38: error: ‘random_access_iterator_tag’ was not declared in this scope
     struct __copy_move<false, false, random_access_iterator_tag>
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by alalazo** [comment]
From [GCC 6 changes](https://gcc.gnu.org/gcc-6/changes.html) : 
> The default mode for C++ is now -std=gnu++14 instead of -std=gnu++98.

I guess a lot of packages are not ready for that...

**Comment by davydden** [comment]
any way to make `tbb`'s concretisation choose another version of `gcc` by default to temporarily fix this issue?

**Comment by tgamblin** [comment]
You can add concretization preferences for this in packages.yaml.

**Comment by tgamblin** [comment]
Might be easier though to just add the gnu++98 flag when `%gcc@6.1:`

**Comment by davydden** [comment]
i tried
```
packages:
  tbb:
    compiler: [gcc@4.8]
```
but the spec still shows `gcc@6.1.0`. I don't think `packages.yaml` is documented at all.

**Comment by tgamblin** [comment]
it's under [Concretization Policies](http://software.llnl.gov/spack/packaging_guide.html#concretization-policies) but should probably be pulled up higher.

**Comment by davydden** [comment]
hm... it does not work as expected:
```
$ spack compiler list
==> Available compilers
-- gcc ----------------------------------------------------------
gcc@6.1.0  gcc@4.8  gcc@4.7  gcc@4.6
davydden@davyddenU1204:~/spack$ spack spec dealii@dev
Input spec
------------------------------
  dealii@dev

Normalized
------------------------------
  dealii@dev
      ^astyle@2.04
      ^blas
      ^bzip2
      ^cmake
      ^lapack
      ^muparser
      ^numdiff
      ^suite-sparse
      ^tbb
      ^zlib

Concretized
------------------------------
  dealii@dev%gcc@6.1.0+arpack~doc+gsl+hdf5+metis+mpi+netcdf+oce+p4est+petsc+slepc+trilinos=linux-x86_64
      ^arpack-ng@3.3.0%gcc@6.1.0+mpi+shared=linux-x86_64
          ^autoconf@2.69%gcc@6.1.0=linux-x86_64
              ^m4@1.4.17%gcc@6.1.0+sigsegv=linux-x86_64
                  ^libsigsegv@2.10%gcc@6.1.0=linux-x86_64
          ^automake@1.15%gcc@6.1.0=linux-x86_64
          ^libtool@2.4.6%gcc@6.1.0=linux-x86_64
          ^openblas@0.2.18%gcc@6.1.0+fpic~openmp+shared=linux-x86_64
          ^openmpi@1.10.2%gcc@6.1.0~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs=linux-x86_64
              ^hwloc@1.11.3%gcc@6.1.0=linux-x86_64
                  ^libpciaccess@0.13.4%gcc@6.1.0=linux-x86_64
      ^astyle@2.04%gcc@6.1.0=linux-x86_64
      ^boost@1.60.0%gcc@6.1.0+atomic+chrono+date_time~debug+filesystem+graph~icu_support+iostreams+locale+log+math+mpi+multithreaded+program_options~python+random+regex+serialization+shared+signals+singlethreaded+system+test+thread+wave=linux-x86_64
          ^bzip2@1.0.6%gcc@6.1.0=linux-x86_64
          ^zlib@1.2.8%gcc@6.1.0=linux-x86_64
      ^cmake@3.5.2%gcc@6.1.0~doc+ncurses+openssl~qt=linux-x86_64
          ^ncurses@6.0%gcc@6.1.0=linux-x86_64
          ^openssl@1.0.2h%gcc@6.1.0=linux-x86_64
      ^gsl@2.1%gcc@6.1.0=linux-x86_64
      ^hdf5@1.8.16%gcc@6.1.0~cxx~debug+fortran+mpi+shared~szip~threadsafe=linux-x86_64
      ^metis@5.1.0%gcc@6.1.0~debug~double~gdb~idx64+shared=linux-x86_64
      ^muparser@2.2.5%gcc@6.1.0=linux-x86_64
      ^netcdf@4.4.0%gcc@6.1.0~hdf4+mpi=linux-x86_64
          ^curl@7.47.1%gcc@6.1.0=linux-x86_64
      ^netcdf-cxx@4.2%gcc@6.1.0=linux-x86_64
      ^numdiff@5.8.1%gcc@6.1.0=linux-x86_64
      ^oce@0.17.1%gcc@6.1.0+tbb=linux-x86_64
          ^tbb@4.4.3%gcc@6.1.0=linux-x86_64
      ^p4est@1.1%gcc@6.1.0+tests=linux-x86_64
          ^lua@5.3.2%gcc@6.1.0=linux-x86_64
              ^readline@6.3%gcc@6.1.0=linux-x86_64
      ^petsc@3.6.3%gcc@6.1.0+boost~complex~debug+double+hdf5+hypre+metis+mpi+mumps+shared+superlu-dist=linux-x86_64
          ^hypre@2.10.1%gcc@6.1.0~internal-superlu+shared=linux-x86_64
          ^mumps@5.0.1%gcc@6.1.0+complex+double+float~idx64~metis+mpi~parmetis~ptscotch~scotch+shared=linux-x86_64
              ^netlib-scalapack@2.0.2%gcc@6.1.0~fpic+shared=linux-x86_64
          ^parmetis@4.0.3%gcc@6.1.0~debug~gdb+shared=linux-x86_64
          ^python@2.7.11%gcc@6.1.0=linux-x86_64
              ^sqlite@3.8.5%gcc@6.1.0=linux-x86_64
          ^superlu-dist@4.3%gcc@6.1.0=linux-x86_64
      ^slepc@3.6.2%gcc@6.1.0~arpack=linux-x86_64
      ^suite-sparse@4.5.1%gcc@6.1.0=linux-x86_64
      ^trilinos@12.6.1%gcc@6.1.0~debug+hdf5+hypre+metis+mumps~python+shared+suite-sparse+superlu-dist=linux-x86_64
          ^glm@0.9.7.1%gcc@6.1.0=linux-x86_64
          ^matio@1.5.2%gcc@6.1.0=linux-x86_64
          ^swig@3.0.8%gcc@6.1.0=linux-x86_64
              ^pcre@8.38%gcc@6.1.0=linux-x86_64
```
with `~/.spack/packages.yaml`
```
packages:
  tbb:
    compiler: [gcc@4.8]
  all:
    compiler: [gcc@6.1.0, gcc]
```

**Comment by davydden** [comment]
p.s. i know why i did not find it, i think it belongs to `basic usage` whereas `Packaging Guide` is supposedly for package developers.

**Comment by Rombur** [comment]
@davydden TBB works fine with gcc-6.1.0 for me. I installed gcc myself and I use the bundled version in deal.II

**Comment by davydden** [comment]
@Rombur thanks for letting me know. I will have a closer look at the errors then...

**Comment by davydden** [comment]
hm, appears to be a `tbb` bug: https://software.intel.com/en-us/forums/intel-threading-building-blocks/topic/610576?language=de

**Comment by davydden** [comment]
this should be fixed in `4.4.4`, i will check if it is the case and if so update the `tbb` package...
